crates/subspace-farmer/src/cluster/nats_client.rs

```diff
@@ -0,0 +1,433 @@
+            Poll::Ready(Some(message)) => match Message::decode(&mut message.payload.as_ref()) {
+                Ok(message) => Poll::Ready(Some(message)),
+                Err(error) => {
+                    warn!(
+                        %error,
+                        message_type = %type_name::<Message>(),
+                        message = %hex::encode(message.payload),
+                        "Failed to decode stream message"
+                    );
+
+                    Poll::Pending
+                }
+            },
+            Poll::Ready(None) => Poll::Ready(None),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+/// NATS client wrapper that can be used to interact with other Subspace-specific clients
+#[derive(Debug, Clone, Deref)]
+pub struct NatsClient {
+    client: Client,
+}
+
+impl From<Client> for NatsClient {
+    fn from(client: Client) -> Self {
+        Self { client }
+    }
+}
+
+impl NatsClient {
+    /// Create new instance by connecting to specified addresses
+    pub async fn new<A: ToServerAddrs>(addrs: A) -> Result<Self, async_nats::Error> {
+        Ok(Self {
+            client: async_nats::connect(addrs).await?,
+        })
+    }
+
+    /// Make request and wait for response
+    pub async fn request<Request>(
+        &self,
+        request: &Request,
+        instance: Option<&str>,
+    ) -> Result<Request::Response, RequestError>
+    where
+        Request: GenericRequest,
+    {
+        let subject = if let Some(instance) = instance {
+            Subject::from(Request::SUBJECT.replace('*', instance))
+        } else {
+            Subject::from_static(Request::SUBJECT)
+        };
+        let message = self
+            .client
+            .request(subject.clone(), request.encode().into())
+            .await?;
+
+        let response =
+            Request::Response::decode(&mut message.payload.as_ref()).map_err(|error| {
+                warn!(
+                    %subject,
+                    %error,
+                    response_type = %type_name::<Request::Response>(),
+                    response = %hex::encode(message.payload),
+                    "Response decoding failed"
+                );
+
+                RequestErrorKind::Other
+            })?;
+
+        Ok(response)
+    }
+
+    /// Make request that expects stream response
+    pub async fn stream_request<Request>(
+        &self,
+        request: Request,
+        instance: Option<&str>,
+    ) -> Result<StreamResponseSubscriber<Request::Response>, StreamRequestError>
+    where
+        Request: GenericStreamRequest,
+    {
+        let stream_request = StreamRequest::new(request);
+
+        let subscriber = self
+            .client
+            .subscribe(stream_request.response_subject.clone())
+            .await?;
+
+        let subject = if let Some(instance) = instance {
+            Subject::from(Request::SUBJECT.replace('*', instance))
```

Comment:
Is '*' part of NATS interface or our implicit wildcard for subject instances?

Reply:
As module documentation says, I strongly recommend to read https://docs.nats.io/nats-concepts/subjects

Reply:
I suspected that it is NATS specific. Consider introducing it as a module constant. Non-critical though if NATS-native code uses it as a standalone char everytime.